### PR TITLE
DDF-3807 Modified logic for finding workspaces in workspace subscriptions

### PR DESCRIPTION
**Pull request by glenhein**
#### What does this PR do?
DDF-3807 Modified logic for finding workspaces in workspace subscriptions to check for subscribed to workspaces first, instead of just all workspaces

Previous logic would query for all workspaces (up to the default query size of 20), and then run them, collect the results, then loop through them to see if any were subscribed to. This is inefficient as it runs workspaces/queries for no reason, and ineffective, as any subscriptions past the first 20 workspaces never get ran.

This logic replaces that, and instead goes through the subscriptions first, up to the first 100, and collects the workspace IDs. It then queries for those IDs instead of all workspaces. The remaining logic stays the same.

#### Who is reviewing it? 
@rzwiefel
@millerw8 
@Bdthomson 
@leo-sakh 

#### Select relevant component teams: 
https://github.com/orgs/codice/teams/ui

#### Ask 2 committers to review/merge the PR and tag them here. If you don't know who to ask, you can request reviews in https://groups.google.com/forum/#!forum/ddf-developers .
(please choose ONLY two committers from below, delete the rest)
@bdeining
@millerw8
#### How should this be tested? (List steps with links to updated documentation)
Set up an environment with many workspaces, and set up email notifications. Subscribe to the workspaces and verify you get email notifications.

#### What are the relevant tickets?
DDF-3807

#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by glenhein** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 1]
```diff
@@ -142,11 +161,11 @@ public String toString() {
   @Override
   public List<WorkspaceMetacardImpl> getWorkspaceMetacards(Set<String> workspaceIds) {
 
-    final List<Filter> filters = createFilters(workspaceIds);
+    if (!workspaceIds.isEmpty()) {
```
Comment (leo-sakh) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: May I suggest inverting the predicate and returning the empty list here early, while getting rid of the larger indented code block?

**Comment by glenhein** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 2]
```diff
@@ -28,60 +27,80 @@
...
-  private final FilterService filterService;
-
   private final SecurityService securityService;
 
+  private final PersistentStore persistentStore;
+
+  private final WorkspaceQueryBuilder workspaceQueryBuilder;
+
+  /** Use an AtomicInteger to make the class thread-safe. */
+  private final AtomicInteger maxSubscriptions = new AtomicInteger(0);
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: If we're only doing sets and gets on this AtomicInteger it means we could just replace it with a volatile variable I believe.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 3]
```diff
@@ -184,24 +213,9 @@ private QueryRequest createQueryRequest(List<Filter> filters) {
...
-    return workspaceIds
-        .stream()
-        .map(
-            id ->
-                filterBuilder.allOf(
-                    filterService.buildMetacardIdFilter(id),
-                    filterService.buildWorkspaceTagFilter()))
-        .collect(Collectors.toList());
+    QueryImpl query = new QueryImpl(filter);
+    query.setPageSize(maxSubscriptions.get());
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: ❓ If this page size is greater than the internal framework limit it means the subscription size could never be greater than that. Did we want to use the ResultIterator here?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 2 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: I check the docs on `volatile`

Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: confirmed, and changed to `volatile`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 4]
```diff
@@ -91,24 +110,24 @@ public String toString() {
...
         + ", catalogFramework="
         + catalogFramework
-        + ", filterBuilder="
-        + filterBuilder
         + ", workspaceTransformer="
         + workspaceTransformer
-        + ", filterService="
-        + filterService
+        + ", maxSubscriptions="
+        + maxSubscriptions.get()
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: ✏️ it would be mighty fine to clean this up into a string format or the like if you're feeling up to it

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 3 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: To be safe, I switched over to using ResultIterable. I followed the approach from the CswEndpoint. Please come back and double-check my impl.

Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Is `query.setPageSize(maxSubscriptions);` still needed when using ResultIterable?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 4 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: done

**Comment by glenhein** [comment]
This PR replaces https://github.com/codice/ddf/pull/3203

**Comment by glenhein** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 5]
```diff
@@ -163,18 +200,27 @@ public String toString() {
...
+      ids =
+          subscriptions
+              .stream()
+              .map(subscription -> subscription.get(ID_FIELD))
+              .filter(Objects::nonNull)
+              .map(Object::toString)
+              .collect(Collectors.toSet());
+    } catch (PersistenceException e) {
+      LOGGER.warn("Failed to get subscriptions for workspaces: {}", e.getMessage());
+      LOGGER.debug("Failed to get subscriptions for workspaces: {}", e.getMessage(), e);
```
Comment (AzGoalie) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: do we need the same message for a warning and a debug log?

File: catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryBuilderImplTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  public void testCreateFilter() {
+
+    String id = "1234";
+
+    FilterBuilder filterBuilder = mock(FilterBuilder.class);
+    FilterService filterService = mock(FilterService.class);
+
+    Filter workspaceTagFilter = mock(Filter.class);
+
+    when(filterService.buildWorkspaceTagFilter()).thenReturn(workspaceTagFilter);
```
Comment (Bdthomson) [inline catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryBuilderImplTest.java]: ✏️ Creating the mocks inline and then doing things with them kind of makes this hard to read in my opinion. I think it would be easier to read if you put all the `= mock(class)` calls at the top of the function (or even use `@Mock` tags and declare at the class level).

File: catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 7]
```diff
@@ -461,15 +461,24 @@
...
+                <argument>
+                    <bean class="org.codice.ddf.catalog.ui.query.monitor.impl.FilterServiceImpl">
+                        <argument ref="filterBuilder"/>
+                    </bean>
+                </argument>
             </bean>
         </argument>
         <argument ref="securityService"/>
+        <argument ref="persistentStore"/>
+        <argument value="100"/>
```
Comment (Bdthomson) [inline catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: If this is configurable via admin UI should there be a `<property>` tag for it also?

File: catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryBuilderImplTest.java (hunk 6 above)
Comment (rzwiefel) [reply catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryBuilderImplTest.java]: Or use spock to be even more readable 😉

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 8]
```diff
@@ -184,24 +230,9 @@ private QueryRequest createQueryRequest(List<Filter> filters) {
...
-  private List<Filter> createFilters(Set<String> workspaceIds) {
-    return workspaceIds
-        .stream()
-        .map(
-            id ->
-                filterBuilder.allOf(
-                    filterService.buildMetacardIdFilter(id),
-                    filterService.buildWorkspaceTagFilter()))
-        .collect(Collectors.toList());
+    QueryImpl query = new QueryImpl(filter);
```
Comment (lamhuy) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: also considering setting a SortBy (primary and secondary sort). if there are more workspaces than maxSubscriptions. The result would be inconsistent and differ for every query.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/api/WorkspaceQueryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.ui.query.monitor.api;
+
+import java.util.Set;
+import org.opengis.filter.Filter;
+
+public interface WorkspaceQueryBuilder {
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/api/WorkspaceQueryBuilder.java]: ❓ Why did we make an interface for this? Looks like we're not registering it as a service anywhere or doing anything else with the interface  (other than creating a bean of the impl and using that)

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rzwiefel) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: One is logging the stacktrace and the other is not. 

Though it does beg the question do we even need a warn? This is not something that a system administrator could resolve.
https://codice.atlassian.net/wiki/spaces/DDF/pages/7045131/Log+Levels

File: catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryBuilderImplTest.java (hunk 6 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryBuilderImplTest.java]: I converted them to `@Mock`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 10]
```diff
@@ -17,103 +17,139 @@
...
+      } catch (CatalogQueryException e) {
+        LOGGER.warn("Error querying for workspaces", e);
+      }
     }
 
     return Collections.emptyList();
   }
 
+  private QueryResponse query(QueryRequest queryRequest) {
+    AtomicLong hitCount = new AtomicLong(0);
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Why use atomic long here in this method?

File: catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 7 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: It is configurable (see `/projects/DdfMaster3/catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.query.monitor.impl.xml`). There doesn't need to be a property tag.

Comment (bdeining) [reply catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: But if it's configurable from the admin ui, how is it updated without a getter / setter?  Do we want to be able to change it after the service is instantiated?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 8 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: done. copied same approach as #3341

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/api/WorkspaceQueryBuilder.java (hunk 9 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/api/WorkspaceQueryBuilder.java]: Mostly out of habit. But there is a test class that plugs in a mock impl.

File: catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImplTest.java [hunk 11]
```diff
@@ -205,18 +182,17 @@ public void testGetWorkspaceMetacardsByWorkspaceIdWithException()
...
+   * @throws PersistenceException
    */
   @SuppressWarnings("unchecked")
   @Test
   public void testGetWorkspaceMetacardsWithException()
-      throws UnsupportedQueryException, SourceUnavailableException, FederationException {
+      throws UnsupportedQueryException, SourceUnavailableException, FederationException,
+          PersistenceException {
 
+    mockCatalogFrameworkQuery("123", "subject");
```
[the remark this answers is not in the document]
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImplTest.java]: done

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 10 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: It's so the hit count can be passed out of the query function, since everything in the query function must be effectively final. This is the same approach taken in `CswEndpoint.java`.

**Comment by glenhein** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 7 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: oh, i see... let me fix that :-)

Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: wait, there is a setter on line 107

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 12]
```diff
@@ -17,103 +17,142 @@
...
 
-    try {
-      return createWorkspaceMetacards(catalogFramework.query(queryRequest));
-    } catch (UnsupportedQueryException | FederationException | SourceUnavailableException e) {
-      LOGGER.warn("Error querying for workspaces", e);
+    if (queryRequest != null) {
+      try {
+        return createWorkspaceMetacards(query(queryRequest));
+      } catch (CatalogQueryException e) {
+        LOGGER.warn("Error querying for workspaces", e);
```
Comment (jrnorth) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Seems like `info` would be more appropriate.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 13]
```diff
@@ -142,17 +181,18 @@ public String toString() {
...
 
-      try {
-        return createWorkspaceMetacards(catalogFramework.query(queryRequest));
-      } catch (UnsupportedQueryException | FederationException | SourceUnavailableException e) {
-        LOGGER.warn("Error querying for workspaces: queryRequest={}", queryRequest, e);
-      }
+    try {
+      return createWorkspaceMetacards(query(queryRequest));
+    } catch (CatalogQueryException e) {
+      LOGGER.warn("Error querying for workspaces: queryRequest={}", queryRequest, e);
```
Comment (jrnorth) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Same here.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 14]
```diff
@@ -17,103 +17,142 @@
...
+          hitCount.compareAndSet(0, sourceResponse.getHits());
+          return sourceResponse;
+        };
+
+    ResultIterable results =
+        ResultIterable.resultIterable(queryFunction, queryRequest, maxSubscriptions);
+
+    List<Result> resultList = results.stream().collect(Collectors.toList());
+
+    long totalHits = hitCount.get();
```
Comment (jrnorth) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Could do this in one line: `final long totalHits = hitCount.get() > 0 ? hitCount.get() : resultList.size();`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 15]
```diff
@@ -163,18 +203,26 @@ public String toString() {
...
    */
-  private QueryRequest createQueryRequestForAllWorkspaceMetacards() {
-    return createQueryRequest(filterService.buildWorkspaceTagFilter());
-  }
+  private QueryRequest createQueryRequestForSubscribedToWorkspaces() {
+    Set<String> ids = new HashSet<>();
+    try {
+      List<Map<String, Object>> subscriptions =
+          persistentStore.get(SubscriptionsPersistentStore.SUBSCRIPTIONS_TYPE);
+      ids =
```
Comment (jrnorth) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: ✏️ This doesn't need to be inside the `try/catch` block. Consider moving the `persistentStore.get()` call and error handling to a separate function so this one can be simplified.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 12 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: changed to `info`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 13 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: changed to `info`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 14 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: changed

File: catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 7 above)
Comment (bdeining) [reply catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: i think it needs to be a property then, so it gets called from blueprint/metatype when the configuration is changed

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 15 above)
Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: refactored

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/api/WorkspaceQueryBuilder.java (hunk 9 above)
Comment (rzwiefel) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/api/WorkspaceQueryBuilder.java]: ❗️ We should definitely remove this then.

Comment (glenhein) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/api/WorkspaceQueryBuilder.java]: the interface has been removed

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 16]
```diff
@@ -17,103 +17,137 @@
...
 
-  private final FilterService filterService;
-
   private final SecurityService securityService;
 
+  private final PersistentStore persistentStore;
+
+  private final WorkspaceQueryBuilder workspaceQueryBuilder;
+
+  /** Use {@code volatile} to make the class thread-safe. */
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: ❓ was the class already threadsafe before the volatile?

Comment (coyotesqrl) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: It's probably a bit of overkill, as it's going to be set by blueprint, but it will ensure that any other threads that _might_ be running queries will see the updated value. We don't generally worry about this, making service properties volatile, but it's not the worst idea.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 17]
```diff
@@ -163,18 +198,30 @@ public String toString() {
...
+  private Set<String> getIdsFromSubscriptions(List<Map<String, Object>> subscriptions) {
+    return subscriptions
+        .stream()
+        .map(subscription -> subscription.get(ID_FIELD))
+        .filter(Objects::nonNull)
+        .map(Object::toString)
+        .collect(Collectors.toSet());
+  }
+
+  private Optional<List<Map<String, Object>>> findSubscriptions() {
```
Comment (jrnorth) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: ✏️ I think it's odd to use `Optional` with collections since you can return an empty collection instead of an empty `Optional`.

File: catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.query.monitor.impl.xml [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+ -->
+<metatype:MetaData xmlns:metatype="http://www.osgi.org/xmlns/metatype/v1.0.0">
+
+    <OCD name="Workspace Service" id="org.codice.ddf.catalog.ui.query.monitor.impl.WorkspaceServiceImpl">
+        <AD id="maxSubscriptions"
+            name="Maximum Subscriptions"
+            description="Specifies the maximum number of subscriptions that may be queried."
```
Comment (jrnorth) [inline catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.query.monitor.impl.xml]: What does this mean to an administrator?

File: catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImplTest.java [hunk 19]
```diff
@@ -40,48 +39,60 @@
...
+  private WorkspaceQueryBuilder workspaceQueryBuilder;
+
   @Before
   public void setup() {
     catalogFramework = mock(CatalogFramework.class);
-    filterBuilder = mock(FilterBuilder.class);
     workspaceTransformer = mock(WorkspaceTransformerImpl.class);
-    filterService = mock(FilterService.class);
     securityService = mock(SecurityService.class);
+    persistentStore = mock(PersistentStore.class);
```
Comment (jrnorth) [inline catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImplTest.java]: There are enough mock objects in here that it's worth using `@Mock` annotations instead.

Comment (Bdthomson) [reply catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImplTest.java]: Fixed.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 17 above)
Comment (Bdthomson) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Changed to 
```java
  private Optional<QueryRequest> createQueryRequestForSubscribedToWorkspaces() {
    final List<Map<String, Object>> subscriptions = findSubscriptions();
    if (subscriptions.isEmpty()) {
      return Optional.empty();
    }
    final Set<String> ids = getIdsFromSubscriptions(subscriptions);
    return Optional.of(getQueryRequestFromIds(ids));
  }
```

at your suggestion.

**Comment by bdeining** [comment]
build now

**Comment by Bdthomson** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 20]
```diff
@@ -14,106 +14,141 @@
...
-    final QueryRequest queryRequest = createQueryRequestForAllWorkspaceMetacards();
-
+  private List<WorkspaceMetacardImpl> queryRequestToWorkspaceMetacards(QueryRequest queryRequest) {
     try {
-      return createWorkspaceMetacards(catalogFramework.query(queryRequest));
-    } catch (UnsupportedQueryException | FederationException | SourceUnavailableException e) {
-      LOGGER.warn("Error querying for workspaces", e);
+      return createWorkspaceMetacards(query(queryRequest));
+    } catch (CatalogQueryException e) {
+      LOGGER.info("Error querying for workspaces", e);
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Why was this changed to info?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 21]
```diff
@@ -142,17 +177,18 @@ public String toString() {
...
 
-      try {
-        return createWorkspaceMetacards(catalogFramework.query(queryRequest));
-      } catch (UnsupportedQueryException | FederationException | SourceUnavailableException e) {
-        LOGGER.warn("Error querying for workspaces: queryRequest={}", queryRequest, e);
-      }
+    try {
+      return createWorkspaceMetacards(query(queryRequest));
+    } catch (CatalogQueryException e) {
+      LOGGER.info("Error querying for workspaces: queryRequest={}", queryRequest, e);
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Should this be at the info level?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java [hunk 22]
```diff
@@ -163,18 +199,35 @@ public String toString() {
...
+        .filter(Objects::nonNull)
+        .map(Object::toString)
+        .collect(Collectors.toSet());
+  }
+
+  private List<Map<String, Object>> findSubscriptions() {
+    try {
+      return persistentStore.get(SUBSCRIPTIONS_TYPE, "", 0, maxSubscriptions);
+    } catch (PersistenceException e) {
+      LOGGER.debug("Failed to get subscriptions for workspaces: {}", e.getMessage(), e);
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Since we are already logging e, it seems redundant to log `e.getMessage()`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 20 above)
Comment (Bdthomson) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: I will change it back.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java (hunk 21 above)
Comment (Bdthomson) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceServiceImpl.java]: Changing to warn.

**Comment by Bdthomson** [comment]
build now

**Comment by Bdthomson** [comment]
build now

**Comment by Bdthomson** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryServiceImpl.java [hunk 23]
```diff
@@ -248,7 +250,7 @@ public void run() {
 
     workspaceTasks
         .stream()
-        .map(ForkJoinPool.commonPool()::submit)
+        .map(new ForkJoinPool(MAX_THREADS)::submit)
```
Comment (stustison) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryServiceImpl.java]: I found the thing I was looking for yesterday: https://github.com/codice/ddf/blob/7c9cb66f3c9b6f1f54686d848f79544559820fcd/catalog/transformer/catalog-transformer-xml/src/main/java/ddf/catalog/transformer/thread/ForkJoinPoolFactory.java
I think you need to use that class. So we need to move that class to somewhere this code can access it. I had similar issues with  the XML transformers and the security manager.

**Comment by Bdthomson** [comment]
build now

**Comment by gjvera** [comment]
🎉 Hero Successful 🎉 
Setup email notifier
set to have a workspace query email subscription
ran wild card search on 20 workspaces after ingesting one item
ran wild card search on 100 workspaces after ingesting one item
got 120 emails

LG2M!
